var/spack/repos/builtin/packages/dray/package.py

```diff
@@ -0,0 +1,332 @@
+    git      = "https://github.com/LLNL/devil_ray.git"
+    url      = "https://github.com/LLNL/devil_ray/releases/download/v0.1.2/dray-v0.1.2.tar.gz"
+
+    maintainers = ['mclarsen', 'cyrush']
+
+    version('develop', branch='develop', submodules='True', preferred=True)
+    version('0.1.2',  sha256='46937f20124b28dc78a634e8e063a3e7a3bbfd9f424ce2680b08417010c376da')
+    version('0.1.1',  sha256='e5daa49ee3367c087f5028dc5a08655298beb318014c6f3f65ef4a08fcbe346c')
+    version('0.1.0',  sha256='8b341138e1069361351e0a94478608c5af479cca76e2f97d556229aed45c0169')
+
+    variant('cuda', default=False, description='Build with CUDA backend')
+    variant('openmp', default=True, description='Build OpenMP backend')
+    variant("shared", default=True, description="Build as shared libs")
+    variant("test", default=True, description='Build unit tests')
+    variant("utils", default=True, description='Build utilities')
+    variant("logging", default=False, description='Enable logging')
+    variant("stats", default=False, description='Enable stats')
+    variant("mpi", default=True, description='Enable MPI compiler')
+
+    depends_on('cuda', when='+cuda')
+    depends_on('mpi', when='+mpi')
+
+    depends_on('cmake@3.9:', type='build')
+    depends_on('cmake@3.14:', when='+cuda', type='build')
+
+    depends_on("conduit~shared", when="~shared")
+    depends_on("conduit+shared", when="+shared")
+
+    depends_on("apcomp~shared+openmp+mpi", when="~shared+openmp+mpi")
+    depends_on("apcomp+shared+openmp+mpi", when="+shared+openmp+mpi")
+    depends_on("apcomp~shared~openmp+mpi", when="~shared~openmp+mpi")
+    depends_on("apcomp+shared~openmp+mpi", when="+shared~openmp+mpi")
+    depends_on("apcomp~shared+openmp~mpi", when="~shared+openmp~mpi")
+    depends_on("apcomp+shared+openmp~mpi", when="+shared+openmp~mpi")
+    depends_on("apcomp~shared~openmp~mpi", when="~shared~openmp~mpi")
+    depends_on("apcomp+shared~openmp~mpi", when="+shared~openmp~mpi")
+
+    depends_on("raja@0.9.0+cuda~openmp+shared", when="+cuda~openmp+shared")
+    depends_on("raja@0.9.0+cuda+openmp+shared", when="+cuda+openmp+shared")
+    depends_on("raja@0.9.0+cuda~openmp~shared", when="+cuda~openmp~shared")
+    depends_on("raja@0.9.0+cuda+openmp~shared", when="+cuda+openmp~shared")
+
+    depends_on("raja@0.9.0~cuda~openmp+shared", when="~cuda~openmp+shared")
+    depends_on("raja@0.9.0~cuda+openmp+shared", when="~cuda+openmp+shared")
+    depends_on("raja@0.9.0~cuda~openmp~shared", when="~cuda~openmp~shared")
+    depends_on("raja@0.9.0~cuda+openmp~shared", when="~cuda+openmp~shared")
+
+    depends_on("umpire@1.0.0+cuda+shared", when="+cuda+shared")
+    depends_on("umpire@1.0.0+cuda~shared", when="+cuda~shared")
+    depends_on("umpire@1.0.0~cuda+shared", when="~cuda+shared")
+    depends_on("umpire@1.0.0~cuda~shared", when="~cuda~shared")
+
+    depends_on("mfem+shared+conduit~threadsafe", when="+shared")
+    depends_on("mfem~shared+conduit~threadsafe", when="~shared")
```

Comment:
This variant forwarding is getting out of hand...

Reply:
I totally agree. This is ascent:
```
###########################################################################
    # package dependencies
    ###########################################################################

    # use cmake 3.14, newest that provides proper cuda support
    # and we have seen errors with cuda in 3.15
    depends_on("cmake@3.14.1:3.14.99", type='build')
    depends_on("conduit~python", when="~python")
    depends_on("conduit+python", when="+python+shared")
    depends_on("conduit~shared~python", when="~shared")
    depends_on("conduit~python~mpi", when="~python~mpi")
    depends_on("conduit+python~mpi", when="+python+shared~mpi")
    depends_on("conduit~shared~python~mpi", when="~shared~mpi")

    #######################
    # Python
    #######################
    # we need a shared version of python b/c linking with static python lib
    # causes duplicate state issues when running compiled python modules.
    depends_on("python+shared", when="+python+shared")
    extends("python", when="+python+shared")
    depends_on("py-numpy", when="+python+shared", type=('build', 'run'))
    depends_on("py-pip", when="+python+shared", type=('build', 'run'))

    #######################
    # MPI
    #######################
    depends_on("mpi", when="+mpi")
    depends_on("py-mpi4py", when="+mpi+python+shared")

    #######################
    # BabelFlow
    #######################
    depends_on('babelflow@develop', when='+babelflow+mpi')
    depends_on('pmt@develop', when='+babelflow+mpi')

    #############################
    # TPLs for Runtime Features
    #############################

    depends_on("vtk-h",             when="+vtkh")
    depends_on("vtk-h~openmp",      when="+vtkh~openmp")
    depends_on("vtk-h+cuda+openmp", when="+vtkh+cuda+openmp")
    depends_on("vtk-h+cuda~openmp", when="+vtkh+cuda~openmp")

    depends_on("vtk-h~shared",             when="~shared+vtkh")
    depends_on("vtk-h~shared~openmp",      when="~shared+vtkh~openmp")
    depends_on("vtk-h~shared+cuda",        when="~shared+vtkh+cuda")
    depends_on("vtk-h~shared+cuda~openmp", when="~shared+vtkh+cuda~openmp")

    # mfem
    depends_on("mfem@4.0.2~threadsafe~openmp+shared+mpi+conduit", when="+shared+mfem+mpi")
    depends_on("mfem@4.0.2~threadsafe~openmp~shared+mpi+conduit", when="~shared+mfem+mpi")

    depends_on("mfem@4.0.2~threadsafe~openmp+shared~mpi+conduit", when="+shared+mfem~mpi")
    depends_on("mfem@4.0.2~threadsafe~openmp~shared~mpi+conduit", when="~shared+mfem~mpi")

    depends_on("adios", when="+adios")

    # devil ray variants wit mpi
    # we have to specify both because mfem makes us
    depends_on("dray@develop+mpi~test~utils+shared+cuda",        when="+dray+mpi+cuda+shared")
    depends_on("dray@develop+mpi~test~utils+shared+openmp",      when="+dray+mpi+openmp+shared")
    depends_on("dray@develop+mpi~test~utils+shared~openmp~cuda", when="+dray+mpi~openmp~cuda+shared")

    depends_on("dray@develop+mpi~test~utils~shared+cuda",        when="+dray+mpi+cuda~shared")
    depends_on("dray@develop+mpi~test~utils~shared+openmp",      when="+dray+mpi+openmp~shared")
    depends_on("dray@develop+mpi~test~utils~shared~openmp~cuda", when="+dray+mpi~openmp~cuda~shared")

    # devil ray variadevelopithout mpi
    depends_on("dray@develop~mpi~test~utils+shared+cuda",        when="+dray~mpi+cuda+shared")
    depends_on("dray@develop~mpi~test~utils+shared+openmp",      when="+dray~mpi+openmp+shared")
    depends_on("dray@develop~mpi~test~utils+shared~openmp~cuda", when="+dray~mpi~openmp~cuda+shared")

    depends_on("dray@develop~mpi~test~utils~shared+cuda",        when="+dray~mpi+cuda~shared")
    depends_on("dray@develop~mpi~test~utils~shared+openmp",      when="+dray~mpi+openmp~shared")
    depends_on("dray@develop~mpi~test~utils~shared~openmp~cuda", when="+dray~mpi~openmp~cuda~shared")


    #######################
    # Documentation related
    #######################
    depends_on("py-sphinx", when="+python+doc", type='build')
    depends_on("py-sphinx-rtd-theme", when="+python+doc", type='build')
```

Reply:
We are almost at the limits of what is practical with variant forwarding, but we aim for things to work close to out of the box.  Building dray with various options does require options in deps.  What is the alternative here? Should we be providing several spack environment files that cover the combos? That also seems hard, are exemplar packages that solve this in a simpler way?

Reply:
I guess the question is how many of these variants are hard requirements? Can we remove the variants from higher packages and rely on lower packages to determine how they should be built? For example, I've seen a lot of packages with a `+foo` variant that doesn't actually do anything, it just specifies how dependencies should be built. But users can already get that control without another variant via:
```console
$ spack install a ^b+foo
```

Reply:
We use all of these variants. Sometimes we try to use global variants, but that breaks down as well.